modules/standard/Random.chpl

```diff
@@ -504,15 +583,92 @@ module Random {
+      of Simple Fast Space-Efficient Statistically Good Algorithms for Random
+      Number Generation" by M.E. O'Neill.
+
+      This class builds upon the :record:`pcg_setseq_64_xsh_rr_32_rng` PCG RNG
+      which has 64 bits of state and 32 bits of output.
+
+      While the PCG RNG used here is believed to have good statistical
+      properties, it is not suitable for cryptographic use. Additionally,
+      if statistical properties of the random numbers are very important,
+      another strategy may be required.
+
+      We have verified that the random numbers generated by this class can match
+      those generated by the C PCG reference implementation. However, this
+      implementation differs from the C PCG reference implementation in how it
+      produces random integers in a range and in support for generating random
+      `real` s.
+
+      Smaller numbers, such as uint(8) or uint(16), are generated from
+      the high-order bits of the 32-bit output.
+
+      To generate larger numbers, several 32-bit-output RNGs are ganged
+      together.  This strategy is recommended by the author of PCG (and
+      demonstrated in the file `pcg32x2-demo.c`. Each of these 32-bit RNGs has a
+      different sequence constant and so will be independent and uncorrelated.
+      For example, to generate 128-bit complex numbers, this RNG will use
+      4 ganged 32-bit PCG RNGs with different sequence constants. One impact of
+      this approach is that this implementation will only generate 2**64
+      different complex numbers (for example).
+
+      This class also supports generating integers within particular bounds.
+      When that is required, this class uses a strategy different from the PCG
+      reference implementation in order to work better a parallel setting. In
+      particular, when more than 1 random value is required as part of
+      generating a value in a range, conceptually it uses more ganged-together
+      RNGs (as with the 32x2 strategy). Each new value beyond the first that
+      is computed will be computed with a different ganged-together RNG.
+      This strategy is meant to avoid statistical bias. While we have tested
+      this strategy to our satisfaction, it has not been subject to rigorous
+      analysis and may have undesireable statistical properties.
+
+      When generating a real, imaginary, or complex number, this implementation
+      uses the strategy of generating a 64-bit unsigned integer and then
+      multiplying it by 2.0**-64 in order to convert it to a floating point
+      number. While this does construct a uniform distribution on rounded
+      floating point values, it leaves out many possible real values (for
+      example, 2**-128). We believe that this strategy has reasonable
+      statistical properties. One side effect of this strategy is that the real
+      number 1.0 can be generated because of rounding. The real number 0.0 can
```

Comment:
We could definitely modify it to generate [0.0, 1.0) but there are two issues:

 1) it wasn't obvious to me that is really better than [0.0, 1.0].
 2) it's not really clear the best way to do that.

The unnecessarily thorough random-floating-point document seemed to prefer [0.0, 1.0] - but NPB for example generates (0.0, 1.0). I suppose it seems easier for a user of these operations to ignore 0.0 or 1.0 if they appear than to add 1.0 back in if it is desired.

The man page for fsetround seems to indicate that it can be expensive to change rounding modes. We could rely on knowledge of IEEE floats and round only the numbers that end up generating 1.0 (the top 1024 uint values) by e.g. setting all of these to a uint we know generates a real < 1.0. That seemed a little bit iffy to me, even though floating point is relatively standardized, I know there can be implementation differences (and e.g. we support turning off exact standard conformance). We could mask and divide by 2**-48 (for example) instead, but then we lose some of the precision that is available.

A related question is whether or not 0.0 should be generated. Apparently some software prefers (0.0, 1.0) because then logarithm is always defined. It seemed obvious to me that 0.0 should be included, but it's not included for NPB for example.

That leaves me wondering if these should be configurable.